File: internal/core/execution.go
// 执行记录数据结构
// ExecutionRecord 是一次运行的完整快照
// 节点状态/日志/变量按调用栈树状组织（RootFrame + Children）
// 同一个集合多次被调用时，按 caller node instance ID 索引区分

package core

import "time"

// ExecutionRecord 单次执行的完整数据
// 通过 Wails 绑定返回给前端时整体序列化为 JSON
type ExecutionRecord struct {
	ID         string            `json:"id"`
	WorkflowID string            `json:"workflow_id"`
	Snapshot   ExecutionSnapshot `json:"snapshot"`     // 启动时的不可变快照
	Status     WorkflowStatus    `json:"status"`       // running / success / failed / terminated
	StartedAt  time.Time         `json:"started_at"`
	FinishedAt *time.Time        `json:"finished_at,omitempty"`
	RootFrame  FrameState        `json:"root_frame"`   // 主流 frame + 嵌套调用的 children
	Error      string            `json:"error,omitempty"`
}

// FrameState 单个调用栈帧的状态（树状递归）
// AssembleID 为空表示主流帧，否则为对应集合 ID
// Children 按 caller node instance ID 索引：同一集合被不同调用节点调用时是不同 frame
type FrameState struct {
	AssembleID string                  `json:"assemble_id"`
	NodeStates map[string]NodeState    `json:"node_states"`
	NodeLogs   map[string][]LogEntry   `json:"node_logs"`
	Variables  map[string]any          `json:"variables"`
	Params     map[string]any          `json:"params,omitempty"`
	Returns    map[string]any          `json:"returns,omitempty"`
	Children   map[string]*FrameState  `json:"children,omitempty"`
}

// ExecutionSnapshot 启动时打的快照
type ExecutionSnapshot struct {
	Workflow  WorkflowDef            `json:"workflow"`
	Assembles map[string]AssembleDef `json:"assembles"` // key = assemble ID
}

// LogEntry 单条节点日志
type LogEntry struct {
	Time    time.Time `json:"time"`
	Level   string    `json:"level"` // info / warn / error
	Message string    `json:"message"`
}

// ExecutionSummary 列表展示用的精简结构（剥离 snapshot / logs / frames）
type ExecutionSummary struct {
	ID           string         `json:"id"`
	WorkflowID   string         `json:"workflow_id"`
	WorkflowName string         `json:"workflow_name"`
	Status       WorkflowStatus `json:"status"`
	StartedAt    time.Time      `json:"started_at"`
	FinishedAt   *time.Time     `json:"finished_at,omitempty"`
	Error        string         `json:"error,omitempty"`
}

// Summary 从 ExecutionRecord 派生 ExecutionSummary
func (r ExecutionRecord) Summary() ExecutionSummary {
	return ExecutionSummary{
		ID:           r.ID,
		WorkflowID:   r.WorkflowID,
		WorkflowName: r.Snapshot.Workflow.Name,
		Status:       r.Status,
		StartedAt:    r.StartedAt,
		FinishedAt:   r.FinishedAt,
		Error:        r.Error,
	}
}

File: frontend/src/types/execution.ts
// 执行记录相关类型，与后端 core.ExecutionRecord 对齐

import type { WorkflowDef, NodeInstance, EdgeConfig } from './workflow';
import type { AssembleDef } from './assemble';

// 工作流执行状态
export type WorkflowStatus =
  | 'Idle'
  | 'Running'
  | 'Success'
  | 'Failed'
  | 'Terminated';

// 节点执行状态
export type NodeState =
  | 'Idle'
  | 'Configuring'
  | 'Checking'
  | 'CheckFailed'
  | 'Ready'
  | 'Executing'
  | 'Success'
  | 'Failed'
  | 'Skipped';

// 日志条目
export interface LogEntry {
  time: string; // ISO 字符串
  level: 'info' | 'warn' | 'error';
  message: string;
}

// 执行启动时打的工作流+集合不可变快照
export interface ExecutionSnapshot {
  workflow: WorkflowDef;
  assembles: Record<string, AssembleDef>;
}

// 完整执行记录
export interface ExecutionRecord {
  id: string;
  workflow_id: string;
  snapshot: ExecutionSnapshot;
  status: WorkflowStatus;
  started_at: string;
  finished_at?: string;
  node_states: Record<string, NodeState>;
  node_logs: Record<string, LogEntry[]>;
  variables: Record<string, unknown>;
  error?: string;
}

// 列表精简结构
export interface ExecutionSummary {
  id: string;
  workflow_id: string;
  workflow_name: string;
  status: WorkflowStatus;
  started_at: string;
  finished_at?: string;
  error?: string;
}

// 便利：从节点快照提取 nodes / edges（执行详情页画布渲染用）
export function snapshotGraph(snapshot: ExecutionSnapshot): {
  nodes: NodeInstance[];
  edges: EdgeConfig[];
} {
  return {
    nodes: snapshot.workflow.nodes,
    edges: snapshot.workflow.edges,
  };
}
